base/c.jl

```diff
@@ -58,11 +58,23 @@ else
     bitstype 32 Cwstring
 end
 
+# construction from typed pointers
 convert{T<:Union{Int8,UInt8}}(::Type{Cstring}, p::Ptr{T}) = box(Cstring, p)
 convert(::Type{Cwstring}, p::Ptr{Cwchar_t}) = box(Cwstring, p)
 convert{T<:Union{Int8,UInt8}}(::Type{Ptr{T}}, p::Cstring) = box(Ptr{T}, p)
 convert(::Type{Ptr{Cwchar_t}}, p::Cwstring) = box(Ptr{Cwchar_t}, p)
 
+# construction from untyped pointers
+convert{T<:Union{Cstring,Cwstring}}(::Type{T}, p::Ptr{Void}) =
+    p==C_NULL ? box(Cstring, p) : throw(ArgumentError("cannot convert non-null void pointer to C string"))
```

Comment:
Then how would `bytestring(Cstring)` behave? Currently, it only checks for `ptr != C_NULL`, relying on the fact that `ptr` points to a NULL-terminated array of chars. I'm not sure what the advantage would be of allowing `Cstring` to contain an arbitrary `Ptr{Void}`

And yes, it only constructs `Cstring`s... Any suggestion? The constructor was based on [@stevengj's comment](https://github.com/JuliaLang/julia/pull/14024#issuecomment-157413181).